File: CppLibrary/Collection/hungary.cpp
#include <iostream>
#include <fstream>
#include <cstring>
using namespace std;
const int limitsize = 201;
int n, m;
bool will[limitsize][limitsize], flag[limitsize];
int c[limitsize];
void init()
{
	memset(will, false, sizeof(will));
	cin >> n >> m;
	int k, x;
	for (int i = 1; i <= n; ++ i)
	{
		fin >> k;
		for (int j = 1; j <= k; ++ j)
		{
			cin >> x;
			will[i][x] = true;
		}
	}	
}

bool extendpath(int x)
{
	for (int i = 1; i <= m; ++ i)
		if (will[x][i] && flag[i])
		{
			flag[i] = false;
			if (c[i] == 0 || extendpath(c[i]))
			{
				c[i] = x;
				return true;
			}
		}
	return false;
}

int hungary()
{
	int ans = 0;
	memset(c, 0, sizeof(c));
	for (int i = 1; i <= n; ++ i)
	{
		memset(flag, true, sizeof(flag));
		if (extendpath(i)) ++ ans;
	}
	return ans;
}

int main()
{
	init();
	cout << hungary() << endl;
	return 0;
}

File: PascalLibrary/Collection/hungary.pas
program hungary;
const
	maxn = 100;
var
	c:array[1..maxn] of longint;
	flag:array[1..maxn] of boolean;
	a:array[1..maxn,1..maxn] of boolean;
	n,m:longint;
	
procedure init;
 var
	p,i,x,y:longint;
 begin
	readln(n,m,p);
	fillchar(c,sizeof(c),0);
	fillchar(a,sizeof(a),false);
	for i:=1 to p do begin
		readln(x,y);
		a[x,y]:=true;
	end;
 end;
 
function extendpath(x:longint):boolean;
 var
	i:longint;
 begin
	for i:=1 to m do
		if (a[x,i]) and (flag[i]) then begin
			flag[i]:=false;
			if (c[i] = 0) or (extendpath(c[i])) then begin
				c[i]:=x;
				exit(true);
			end;
		end;
	extendpath:=false;
 end;
 
procedure main;
 var
	i,ans:longint;
 begin
	ans:=0;
	for i:=1 to n do begin
		fillchar(flag,sizeof(flag),true);
		if extendpath(i) then inc(ans);
	end;
	writeln(ans);
 end;
 
begin
	init;
	main;
end.
